src/main.rs

```diff
@@ -0,0 +1,291 @@
+                 led1_ref: &'static mut LED1,
+                 led3_ref: &'static mut LED3,
+                 button1_ref: &'static mut BUTTON1) {
+    let mut msg : String<64> = String::new();
+    msg.clear();
+    core::fmt::write(&mut msg, format_args!("Boot\n")).unwrap();
+    _ = (uart_ref).write(msg.as_bytes()).await;
+    let mut hr = Hr::new();
+    let mut n0 = 0usize;
+    let mut count0 = 0u32;
+    loop {
+        let sample = SAMPLE_SIGNAL.wait().await;
+        let lp = button1_ref.get_level() == Level::Low;
+        led3_ref.set_level(if !lp { High } else { Low });
+        let (n, cooked_sample, peak, state) = hr.tick(lp, sample);
+        led1_ref.set_level(if state != 0 { High } else { Low });
+        if DUMP_MODE {
+            msg.clear();
+            // core::fmt::write(&mut msg, format_args!("{} {}\n",  sample, state+peak)).unwrap();
+            core::fmt::write(&mut msg, format_args!("{} {}\n",  cooked_sample, if lp {1} else {0})).unwrap();
+            // core::fmt::write(&mut msg, format_args!("{}\n",  sample)).unwrap();
+            _ = (uart_ref).write(msg.as_bytes()).await;
+            continue;
+        }
+        if peak != 0 {
+            let rate = hr.hr();
+            let (a, b) = hr.above_below();
+            let d = a-b;
+            msg.clear();
+            let count = c5412::get_count();
+            let refresh = 1000f64 * (count-count0) as f64 / (n-n0) as f64;
+            core::fmt::write(&mut msg, format_args!("{} {} rate={:.2} refresh={:.2}\n",
+                                                    n-n0, d, rate, refresh)).unwrap();
+            _ = (uart_ref).write(msg.as_bytes()).await;
+            count0 = count;
+            n0 = n;
+        } else if n-n0 > 3000 {
+            let (dc, thresh) = hr.help();
+            msg.clear();
+            core::fmt::write(&mut msg, format_args!("Help: {} {}\n",  dc, thresh)).unwrap();
+            _ = uart_ref.write(msg.as_bytes()).await;
+            n0 = n;
+        }
+    }
+}
+
+#[embassy_executor::main]
+async fn main(spawner: Spawner) {
+    let mut p = embassy_stm32::init(Default::default());
+    let button1 = Input::new(p.PC13, Pull::None);
+    let led1 = Output::new(p.PB0, Level::High, Speed::Low);
+    let led3 = Output::new(p.PB14, Level::High, Speed::Low);
+    let uart = UartTx::new(p.USART3, p.PD8, p.DMA1_CH1, Config::default()).unwrap();
+    let mut delay = Delay;
+    let mut adc = Adc::new(p.ADC1, &mut delay);
+
+    // Turn on ADC oversampling
+    // 0x4002200c: 0x80000008
+    // 0x40022010: 0x000f0001 // f=16x oversample, 001 = oversample on
+    unsafe { let p : *mut u32 = 0x4002200c as *mut u32; *p = 0x80000008; } // 008=12 bit
+    unsafe { let p : *mut u32 = 0x40022010 as *mut u32; *p = 0x000f0001; } // f=16x oversample, 001=ovs on
+    // Turn down clock
+    unsafe { let p : *mut u32 = 0x40022308 as *mut u32; *p = 6 << 18; } // Slow down clock to /12 (0x0018)
+
+    let uart_ref = UART_INST.init(uart);
```

Comment:
Possibly combine all the related but dissimilar objects into clusters
* uart = UartTx::new
* uart_ref = UART_INST.init(uart)

etc

Reply:
done